### PR TITLE
981 filter by location

### LINKED ISSUE #981: As any user  I want to filter events by location, so that I can select the most acceptable to me
**As any** user   **I want to** filter events by location, **so that** I can select the most acceptable to me

**Date Time**
12-28-2021

**Pre-requisites**
The event creation functional is implemented

**Description**
_Common filter charactaristics_
1. Any user can use filters functionality which is described below: 
2. Clicking on the arrow button on the right side of the Home page opens a modal window with filters   
image
_Note:_ by closing the modal window all the chosen filters should be kept by the system
 3. Modal window include next filters: 
 - "Category"- field type "menu tree" (will be defined in another US [link]) 
 - "Organizer" - field type "search field"(will be defined in another US [link]) 
 - "Location" - field type "menu tree" (will be defined in another US [link]) 
 - "Age" - field type "menu tree" (will be defined in another US [link]) 
 - "Calendar"- field type calendar\period picker (will be defined in another US [link])
4. Clicking on filter name\header - expand or collapse tree with a list of filter's options 
5. By default: 
- all filters don't have preselected values (except calendar, it has predefined start date = today) 
- Apply button is disabled (after selecting\changing values in filters Apply button to become enabled)
6. Clicking on Apply button (when it is enabled) - apply filtering criteria to a list of events and visualize only those which match the criteria 
_Note_: after applying filters modal windows should stay open 
7. Field type "menu tree": 
- Clicking on the checkbox - select deselect value\item\option\criteria 
- Selected value marked with a green stick 
- Deselected value is with an empty square 
- Multiple options can be selected simultaneously Note: when all values\items\options\criterias are deselected it is the same when they all are selected
 8. Clicking on the "Clear" button return all filters to their default states 
 9. When the user wants to close the modal window with filters, he has to click on the arrow
image
10. If the selection of events is empty when applying the filter criteria, the text: "No events meet the specified criteria. Please make another choice."  is displayed on the home page.

_Location filter_
11. Any user can use the functionality of filter "Location" which is defined below:
- field type - "menu tree" with single selection (User can select only one value "Near me" or "Online") and button 'Choose on the map';
- default value - empty;
- Behavior: Clicking on checkbox - select deselect value\item\option\criteria
- Selected value marked with a green stick
- Deselected value is with an empty square
- Click on the button 'Choose on the map' provides the possibility to mark the location of the event on the map implemented in the modal window. After the selection of the geographic point, the events would be filtered due to the location within a radius of 10 km

_Note: after applying filters modal window should stay open_

**Acceptance criteria**
[Given] any user on the main page on the modal window with filters
[When] the user clicks options in the tree menu of a location filter 
[Then] the group with checkboxes and a button is opened
[AND]
[Given] any user on filter group location
[When] the user checks in the option 'near me'
[Then] the system filter events on the home page due to the events' location within a radius of 10 km around the location of the user
[and] 
[Then] checkbox 'Online' and button 'Choose on the map'  are disabled
[OR]
[Given] any user on filter group location
[When] the user checks in the option 'online'
[Then] the system filter events on the home page due to the online events
[and] 
[Then] checkbox 'near me' and button 'Choose on the map'  are disabled
[OR]
[Given] any user on filter group location
[When] the user clicks the button  'Choose on the map' 
[Then] the modal window with the map is opened
[AND]
[When] the user puts the mark point and press 'Apply'
[Then] the system filter events on the home page due to the events' location within a radius of 10 km around the mark pointed by the user
[and] 
[Then] checkboxes 'near me' and 'Online'  are disabled

**Additional context**
image
image

**Epic link** Epic 12 #934 US #979 

**Labels to be added** "User story", Corresponding Functionality (e.g. User), Priority ("pri: "), "Corresponding Epic"

### PR DESCRIPTION
**Pull request by rusnakvlad**
dev
## GitHub Project

* [Main GitHub Project ticket](https://github.com/EventsExpress/EventsExpress/projects/2)

## Code reviewers

- [ ] @iuriikhrystiuk 
- [ ] @sand0 

### Second Level Review

- [ ] @idontwannawakeup 

## Summary of issue

Was the requirement to implement filtering by location.

## Summary of change

Was implemented filtering by location.

## Testing approach

ToDo

## CHECK LIST
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: EventsExpress/ClientApp/src/actions/events/filter/actions.js [hunk 1]
```diff
@@ -1,14 +1,22 @@
 import { stringify, exclude } from 'query-string';
 import { get_events } from '../../event/event-list-action';
+import { enumLocationType } from '../../../constants/EventLocationType';
 
 export const applyFilters = filters => {
     return async dispatch => {
         filters.owners = filters?.organizers?.map(organizer => organizer.id);
-
+        filters.locationtype = filters?.location.type;
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/actions/events/filter/actions.js:8]: Are you sure, that it should be `locationtype` not `locationType`? Would this property be bound properly?

File: EventsExpress/ClientApp/src/components/event/filter/parts/location/location-filter.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import { LocationMapWithCircle } from "../../../../helpers/form-helpers/location";
+import DisplayMap from "../../../../event/map/display-map";
+import "../../../slider.css";
+const LocationFilter = ({ dispatch, location, formValues }) => {
+  const clear = () => dispatch(change("filter-form", "location", {type: null} ));
+
+  const onChangeLocationType = (event) => {
+    const type = Number(event.target.value);
+    if (type === enumLocationType.map) {
+      dispatch(
```
Comment (iuriikhrystiuk) [inline EventsExpress/ClientApp/src/components/event/filter/parts/location/location-filter.js]: Please use mapDispatchToProps and do not call dispatch inside components directly.

**Review by iuriikhrystiuk — COMMENTED** [review]
Did you remove the old location filter?

File: EventsExpress/ClientApp/src/actions/events/filter/actions.js (hunk 1 above)
Comment (rusnakvlad) [reply EventsExpress/ClientApp/src/actions/events/filter/actions.js:8]: yes it works so (locationtype)

File: EventsExpress/ClientApp/src/components/event/filter/parts/location/location-filter.js (hunk 2 above)
Comment (rusnakvlad) [reply EventsExpress/ClientApp/src/components/event/filter/parts/location/location-filter.js]: Done

**Review by iuriikhrystiuk — APPROVED** [review]

**Pull request merged**